homeassistant/components/vizio/media_player.py

```diff
@@ -52,34 +57,56 @@
 PLATFORM_SCHEMA = vol.All(PLATFORM_SCHEMA.extend(VIZIO_SCHEMA), validate_auth)
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+async def async_setup_platform(
+    hass: HomeAssistantType,
+    config: ConfigType,
+    async_add_entities: Callable[[List[Entity], bool], None],
+    discovery_info: Dict[str, Any] = None,
+):
     """Set up the Vizio media player platform."""
+
     host = config[CONF_HOST]
     token = config.get(CONF_ACCESS_TOKEN)
     name = config[CONF_NAME]
     volume_step = config[CONF_VOLUME_STEP]
     device_type = config[CONF_DEVICE_CLASS]
-    device = VizioDevice(host, token, name, volume_step, device_type)
-    if not device.validate_setup():
+
+    device = VizioAsync(
+        DEVICE_ID, host, name, token, device_type, async_get_clientsession(hass, False)
+    )
+    if await device.can_connect():
+        async_add_entities(
+            [VizioDevice(hass, device, host, token, name, volume_step, device_type)],
+            True,
+        )
+    else:
         fail_auth_msg = ""
         if token:
-            fail_auth_msg = " and auth token is correct"
+            fail_auth_msg = ", auth token is correct"
         _LOGGER.error(
             "Failed to set up Vizio platform, please check if host "
-            "is valid and available%s",
+            "is valid and available, device type is correct%s",
             fail_auth_msg,
         )
         return
 
-    add_entities([device], True)
-
 
 class VizioDevice(MediaPlayerDevice):
     """Media Player implementation which performs REST requests to device."""
 
-    def __init__(self, host, token, name, volume_step, device_type):
+    def __init__(
+        self,
+        hass: HomeAssistantType,
```

Comment:
Don't pass in `hass`. It will be set on the entity when it has been added to home assistant.